Previously "flow" for operator record was forbidden, but defined! (#3094)
modelica/ModelicaSpecification: chapters/classes.tex

A \firstuse{prefix} is property of an element of a class definition which can be present or not be present, e.g., \lstinline!final!, \lstinline!public!, \lstinline!flow!.

> Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type,

Author: I have tried to add a clarification that not all of those combinations are legal.

> Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type,

Author: > I looked through the code and the test cases I have because I remember implementing the support the way it is described in the paragraph on line 230 (master line numbering). If I change that to not allow flows and streams to be records (as required by the suggested lines 228 and 229) then I get a few test failures. But all of them are the tests that we wrote to test the implementation of this paragraph. So I could not find any real life example. However, how can we be sure that there are no user models that would use that feature?

My logic was that for `flow `the specification previously said:

> Variables declared with the flow or the stream type prefix shall be a subtype of Real

However, we could change it so that 'flow' is applied to elements of structured components, except for operator record and allow that. That would be backward compatible even for those tests.

For `stream `there's the additional requirement that it has to be a scalar Real for the rest of the stream-handling logic; having an exception for a record with a scalar Real just seems contrived - and I don't think it has been implemented.

> Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type,

Reviewer: Sorry, last time I had to rush into a meeting and had time to summarize the problem but no time to offer a suggestion. I think with the requirements that connection sets are build up from primitive variables (as defined in Generation of Connect Equation), it is still valid to have structured components declared as flow and streams, as long as the primitive variables are subtype of Real or operator records for flows and subtype of Real for streams. And the additional requirement on flows when there ares streams present. At least, this, I believe, would be enough to fit with other requirements of the specification without contradictions or making the change backward-incompatible. I don't know if it would make sense from engineering/physical point of view to have records of streams. But it does from programmatic point of view.

So, I think, instead of the changes to classes.tex that have been applied so far in this PR, we need to remove the paragraph:
> Variables declared with the flow or the stream type prefix shall be a subtype of Real.

and insert after the paragraph that explains application of prefixes to structured components (line 230 of master):
> The primitive elements with prefix \lstinline!stream! shall be a subtype of Real.
> The primitive elements with prefix \lstinline!flow! shall be a subtype of Real or an operator record component.

Then we should probably make the definition of the primitive elements in Generation of Connect Equation:
> The primitive elements are of simple types or of types defined as operator record (i.e. a component of an operator record type is not split into sub-components).

into a proper definition, so that we can reference it in the suggested text above. The change to streams.tex as suggested in this PR should also be applied. And the paragraph on line 230 should be split into one sentence per line.

> Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type,

Author: I will change along those lines for flow; especially as you can have a record with multiple elements.
But not for stream as there are so many formulations that would need to be changed - and since there can anyway only be one scalar stream in a connector, it would only allow a record with one scalar Real element.

> Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type,

Author: I have now done that, is it ok?

> record, operator record, and connector components -- see also record specialized class, \cref{specialized-classes}.

Reviewer 2: Remove line-break inside sentence.  (Unable to prepare suggestion.)

Author: Have done.

An exception is \lstinline!input! for components whose type is of the special class function type (these can only be used for function formal parameters and has special semantics, see \cref{functional-input-arguments-to-functions}), and the \lstinline!input! prefix is not applied to the elements of the component and is allowed even if the elements have \lstinline!input! or \lstinline!output! prefix.

In addition, instances of classes extending from \lstinline!ExternalObject! may have type prefixes \lstinline!parameter! and \lstinline!constant!, and in functions also type prefixes
\lstinline!input! and \lstinline!output!, see \cref{external-objects}.

Variables declared with the \lstinline!stream! type prefix shall be a subtype of \lstinline!Real!.

> Variables declared with the \lstinline!flow! type prefix shall be a subtype of \lstinline!Real! or an operator record definining certain operators, see \cref{generation-of-connection-equations}.

Reviewer 2: Wouldn't it be nice to mention that this is related to the need for an additive group?  For instance,
```suggestion
Variables declared with the \lstinline!flow! type prefix shall be a subtype of \lstinline!Real! or an operator record defining an additive group, see \cref{generation-of-connection-equations}.
```

This is further restricted in \cref{definition-of-stream-connectors}.

The type prefixes \lstinline!input! and \lstinline!output! of a structured component (except as described above) are also applied to the elements of the component (this is done after verifying that the type prefixes occurring on elements of the component are correct).
When any of the type prefixes \lstinline!input! and \lstinline!output! are applied for a structured component, no element of the component may have any of these type prefixes, nor can they have \lstinline!flow! or \lstinline!stream! prefixes.
The corresponding rules for the type prefixes \lstinline!discrete!, \lstinline!parameter! and \lstinline!constant! are described in \cref{variability-of-structured-entities} for structured components.

> \begin{nonnormative}

Reviewer 2: Empty lines on both sides of `nonnormative` tend to make the source more readable:
```suggestion

\begin{nonnormative}
```

> The prefixes \lstinline!flow! and \lstinline!stream! are missing from this list, as the corresponding rules would not be relevant based on other semantic restrictions.

Reviewer 2: There are many things that could be interpreted as _this list_.  Some clarification is needed.

Author: ok

A \firstuse{prefix} is property of an element of a class definition which can be present or not be present, e.g., \lstinline!final!, \lstinline!public!, \lstinline!flow!.

> Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type, record, operator record, and connector components -- see also record specialized class, \cref{specialized-classes}.

Reviewer: Does it makes sense to include stream in this list at all? To me, the restriction on stream on line 228 is enough.

Author: Since we have type-prefix in the grammar I think it makes sense that this definition of type prefixes matches that.

This is further restricted below; some of these combinations of type prefixes and specialized classes are not legal.

An exception is \lstinline!input! for components whose type is of the special class function type (these can only be used for function formal parameters and has special semantics, see \cref{functional-input-arguments-to-functions}), and the \lstinline!input! prefix is not applied to the elements of the component and is allowed even if the elements have \lstinline!input! or \lstinline!output! prefix.

In addition, instances of classes extending from \lstinline!ExternalObject! may have type prefixes \lstinline!parameter! and \lstinline!constant!, and in functions also type prefixes

> \lstinline!input! and \lstinline!output!, see \cref{external-objects}.

Reviewer: Can we fix the line break here?

Author: I was unsure what you wanted - I added a line-break above. If that wasn't correct, please explain.

Reviewer: Sorry, that was not clear. I meant the part on line 227 is actually the same sentence that started on line 226. According to the guidelines, we should have one sentence on one line. And this sentence seems to have not been updated. Unless there is some reason we kept it on two lines? That's not really part of this PR but while we are changing this chapter and while I noticed it, I thought it might be ok to change it here instead of opening a separate PR.

Author: Ok, I undid the previous one and merged that line.

Variables declared with the \lstinline!stream! type prefix shall be a subtype of \lstinline!Real!.
This is further restricted in \cref{definition-of-stream-connectors}.

> The type prefix \lstinline!flow! of a component that is not a primitive element (see \cref{primitive-elements}), is also applied to the elements of the component (this is done after verifying that the type prefixes occurring on elements of the component are correct).

Reviewer: What is meant by "the type prefixes occurring on elements of the components are correct"?

Author: It was present before this PR, and to me it indicates that we first check that you don't have an invalid type prefix on the component - before "overriding" it.

Reviewer: It would make sense to me that the "correct" would mean satisfying the restriction that is now on line 237:
>When any of the type prefixes \lstinline!flow!, \lstinline!input! and \lstinline!output! are applied for a structured component, no element of the component may have any of these type prefixes, nor can they have \lstinline!stream! prefix.

Are there any other restrictions? I find it confusing that we refer to the restrictions as if they are already known. But if we are talking about restrictions on line 237 we haven't introduced them yet.

But, you are right, this has existed before this PR. So maybe we need to open a separate issue or PR for it?

Reviewer: I tried to create a ticket for the issue but the text is changed so much by this PR, it is hard to base it off the master the way it is now. So I'm approving this PR now and will create a ticket once it is merged.

Primitive elements with the \lstinline!flow! type prefix shall be a subtype of \lstinline!Real! or an operator record defining an additive group, see \cref{generation-of-connection-equations}.

The type prefixes \lstinline!input! and \lstinline!output! of a structured component (except as described above) are also applied to the elements of the component (this is done after verifying that the type prefixes occurring on elements of the component are correct).

When any of the type prefixes \lstinline!flow!, \lstinline!input! and \lstinline!output! are applied for a structured component, no element of the component may have any of these type prefixes, nor can they have \lstinline!stream! prefix.
The corresponding rules for the type prefixes \lstinline!discrete!, \lstinline!parameter! and \lstinline!constant! are described in \cref{variability-of-structured-entities} for structured components.

\begin{nonnormative}

> The prefixes \lstinline!stream! is not included with \lstinline!flow!, \lstinline!input! and \lstinline!output!, as the corresponding rules would not be relevant based on other semantic restrictions.

Reviewer: Should be singular:
```suggestion
The prefix \lstinline!stream! is not included with \lstinline!flow!, \lstinline!input! and \lstinline!output!, as the corresponding rules would not be relevant based on other semantic restrictions.
```
Does it still make sense to include it? I think if stream is removed from the list on line 220 as I commented above, it would not be necessary to have this sentence or the one below. It would also be easier to follow.

Also it's still not very clear where stream is not included.

Author: I rewrote that - if it still not clear, I would propose to just remove that entire non-normative block.

To me it makes sense to explain that stream, flow, input, output could be treated uniformly - if we ignore semantically invalid combinations; but it is just intended to help - so if it doesn't it is better to remove it.

A \firstuse{prefix} is property of an element of a class definition which can be present or not be present, e.g., \lstinline!final!, \lstinline!public!, \lstinline!flow!.

Variables declared with the \lstinline!stream! type prefix shall be a subtype of \lstinline!Real!.

> Variables declared with the \lstinline!flow! type prefix shall be a subtype of \lstinline!Real! or an operator record definining certain operators, see \cref{generation-of-connection-equations}.

Reviewer: The sentence on line 224 allows streams and flows to be applied to record and connector components. (Sorry, GitHub does not allow me to add the comment to that line.) It does not mention operator record explicitly. Is it assumed that record means operator record? Either way, there is still a contradiction between lines 220-221 and 224.

Author: I agree it is a mess, and I have now tried to improve it.
Some of the problem was that I missed some text on ridiculously long lines.

Note that the text for applying stream, flow, input, output the text is somewhat more complicated; since even if we don't allow 'stream' on a structured component we still want to keep the restriction that if you use 'input' on a structured component there are no 'stream' members in it.

That section could still be cleaned up/restructured further, but I think that can wait.
